### PR TITLE
Add initial progress on catamorphisms

### PR DESCRIPTION
**Pull request by paulyoung**
cc @paf31

### CODE REVIEW

File: src/WebIDLFFI.purs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  let lines = nodeToPs <$> nodes
+  return $ "module " ++ moduleName ++ " where\n\n" ++ intercalate "\n" lines ++ "\n"
+  -- return $ show nodes
+
+  where
+
+  firstToLower :: String -> String
+  firstToLower s = case uncons s of
+    Just { head: x, tail: xs } -> fromChar (toLower x) ++ xs
+    Nothing -> ""
```
Comment (paulyoung) [inline src/WebIDLFFI.purs]: I suppose this should be `s` instead of `""`.

File: src/WebIDLFFI.purs [hunk 2]
```diff
@@ -0,0 +1,151 @@
+module WebIDLFFI where
```
Comment (paf31) [inline src/WebIDLFFI.purs]: Prefer explicit exports if possible:

```purescript
module WebIDLFFI (toPureScript) where
```

File: src/WebIDLFFI.purs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+import Data.Foldable (intercalate)
+import Data.Maybe (Maybe(Nothing, Just))
+import Data.String (fromChar, uncons)
+import WebIDL (Argument, Member(OtherMember, FieldMember, ConstantMember, AttributeMember, OperationMember), Node(InterfaceNode, OtherNode, EnumNode, ExceptionNode, DictionaryNode, CallbackNode, TypeDefNode, ImplementsNode), Type(NullableType, GenericType, NamedType, ArrayType, UnionType), parse)
+
+toPureScript' :: forall eff. String -> String -> Eff (err :: EXCEPTION | eff) String
+toPureScript' moduleName idl = do
+  nodes <- parse idl
+  let lines = nodeToPs <$> nodes
+  return $ "module " ++ moduleName ++ " where\n\n" ++ intercalate "\n" lines ++ "\n"
```
Comment (paf31) [inline src/WebIDLFFI.purs]: Consider moving the intercalate outside:

```purescript
return $ intercalate "\n"
  [ "module " ++ moduleName ++ " where"
  , ""
  ] <> lines
```

File: src/WebIDLFFI.purs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  node f _ _ _ _ _ _ _ (InterfaceNode x) = f x
+  node _ f _ _ _ _ _ _ (ImplementsNode x) = f x
+  node _ _ f _ _ _ _ _ (TypeDefNode x) = f x
+  node _ _ _ f _ _ _ _ (CallbackNode x) = f x
+  node _ _ _ _ f _ _ _ (DictionaryNode x) = f x
+  node _ _ _ _ _ f _ _ (ExceptionNode x) = f x
+  node _ _ _ _ _ _ f _ (EnumNode x) = f x
+  node _ _ _ _ _ _ _ f (OtherNode x) = f x
+
+  interfaceNodeToPs :: { name :: String, partial :: Boolean, members :: Array Member, inheritance :: Maybe String } -> String
```
Comment (paf31) [inline src/WebIDLFFI.purs]: Consider using open rows if you don't need all of the labels here:

```purescript
interfaceNodeToPs :: forall r. { name :: String, members :: Array Member | r } -> String
```

File: src/WebIDLFFI.purs [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  enumNodeToPs x = "FIXME"
+
+  otherNodeToPs :: String -> String
+  otherNodeToPs x = "FIXME"
+
+  nodeToPs :: Node -> String
+  nodeToPs = node interfaceNodeToPs implementsNodeToPs typeDefNodeToPs callbackNodeToPs dictionaryNodeToPs exceptionNodeToPs enumNodeToPs otherNodeToPs
+
+toPureScript :: String -> String -> Either Error String
+toPureScript moduleName idl = runPure (catchException (return <<< Left) (Right <$> toPureScript' moduleName idl))
```
Comment (paf31) [inline src/WebIDLFFI.purs]: Perhaps pull this out into a `where` clause:

```purescript
toPureScript moduleName idl = reifyErrors $ toPureScript' moduleName idl
  where
    reifyErrors :: forall a. (forall eff. Eff (err :: EXCEPTION | eff) a) -> Either Error a
    reifyErrors x = runPure (catchException (return <<< Left) (map Right x))
```

**Comment by paf31** [comment]
Looks great so far! Thanks for working on this.

**Comment by paulyoung** [comment]
Thanks for the feedback!

So you think the general approach is OK? I was heading towards using a fold but would love to know if you have any other ideas.

**Comment by paf31** [comment]
I think a fold should work just fine. I can't think of any cases where your generated code would depend on any larger context.

**Comment by paulyoung** [comment]
OK. Let me know if you have any other feedback, even if it feels like nitpicking.

File: src/WebIDLFFI.purs (hunk 2 above)
Comment (paulyoung) [inline src/WebIDLFFI.purs]: Is it possible to configure implicit exports as a warning?

File: src/WebIDLFFI.purs (hunk 3 above)
Comment (paulyoung) [inline src/WebIDLFFI.purs]: Looks like it needs to be something like:

```diff
-return $ intercalate "\n"
+return $ intercalate "\n" $
     [ "module " ++ moduleName ++ " where"
     , ""
-    ] <> lines
+    ] <> lines <> [ "" ]
```

or 

```diff
-return $ intercalate "\n"
+return $ intercalate "\n" $
     [ "module " ++ moduleName ++ " where"
     , ""
-    ] <> lines
+    ] <> snoc lines ""
```

Any preference or other suggestions?

File: src/WebIDLFFI.purs (hunk 4 above)
Comment (paulyoung) [inline src/WebIDLFFI.purs]: Good call. I just copied the type signature and didn't think much about it.

I'm sort of used to the Swift compiler and eslint warning about unused values.

File: src/WebIDLFFI.purs (hunk 5 above)
Comment (paulyoung) [inline src/WebIDLFFI.purs]: If I do this, I get the following error:

```
[1/1 EscapedSkolem] src/WebIDLFFI.purs:153:1

  153  toPureScript :: String -> String -> Either Error String
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  A type variable has escaped its scope.

  in the expression ($) reifyErrors
  in value declaration toPureScript
```

I don't get the error if I don't provide a type signature for `reifyErrors`. What would you suggest?

**Comment by paulyoung** [comment]
I noticed a difference between our indentation and line spacing when using `where`.

Is your way the convention?

File: src/WebIDLFFI.purs (hunk 5 above)
Comment (paf31) [inline src/WebIDLFFI.purs]: Try this:

```purescript
toPureScript moduleName idl = reifyErrors (toPureScript' moduleName idl)
  where
    reifyErrors :: forall a. (forall eff. Eff (err :: EXCEPTION | eff) a) -> Either Error a
    reifyErrors x = runPure (catchException (return <<< Left) (map Right x))
```

**Comment by paf31** [comment]
I don't think there's a convention really. I tend to use that way for Haskell and PureScript to try to make the scope of things clearer (IMO).

File: src/WebIDLFFI.purs (hunk 5 above)
Comment (paulyoung) [inline src/WebIDLFFI.purs]: That works. Why does using parentheses instead of `$` make a difference?

Comment (paf31) [inline src/WebIDLFFI.purs]: It's a typing restriction due to impredicativity. This explains it pretty well: http://stackoverflow.com/questions/9468963/runst-and-function-composition

**Pull request merged**